### PR TITLE
Enhance comparing tags and fields in the where clause

### LINKED ISSUE #3371: query with WHERE tag=field doesn't work
I'm not sure what the intended behavior is, but the query engine won't allow me to use a field value as a matcher for a tag value, and vice versa.

Insert a point and verify it's there.

```
> insert tag_eq_field,a=foo value="foo"
> select * from tag_eq_field
name: tag_eq_field
tags: a=foo
time				value
----				-----
2015-07-17T20:53:02.653672103Z	foo
```

I can query the point by the tag or the field using the right string.

```
> select * from tag_eq_field where value='foo'
name: tag_eq_field
tags: a=foo
time				value
----				-----
2015-07-17T20:53:02.653672103Z	foo

> select * from tag_eq_field where a='foo'
name: tag_eq_field
tags: a=foo
time				value
----				-----
2015-07-17T20:53:02.653672103Z	foo
```

However, querying the point with tag=field or vice versa returns no points, and they have different behaviors from each other:

```
> select * from tag_eq_field where a=value
> select * from tag_eq_field where value=a
name: tag_eq_field
tags: a=foo
time	value
----	-----

> 
```

### PR DESCRIPTION
**Pull request by jsternberg**
Now it is possible to compare tags and fields and it is also now
possible to compare tags and tags. Previously, it was only possible to
compare fields with fields and tags with a string or a regex.

Fixes #3371.

### CODE REVIEW

**Comment by jsternberg** [comment]
@benbjohnson

File: tsdb/engine/tsm1/iterator.gen.go.tmpl [hunk 1]
```diff
@@ -108,7 +108,7 @@ type {{.name}}Iterator struct {
 	aux   []cursorAt
 	conds struct {
 		names []string
-		curs  []*bufCursor
+		curs  []cursorAt
```
Comment (benbjohnson) [inline tsdb/engine/tsm1/iterator.gen.go.tmpl]: Why is this changed to `cursorAt`?

Comment (jsternberg) [inline tsdb/engine/tsm1/iterator.gen.go.tmpl]: This is so conditions can include tag values too. If a condition only has tags, they get filtered out. If a condition has a field and a tag, then we need a fake cursor for the tag to compare it with the field.

Comment (benbjohnson) [inline tsdb/engine/tsm1/iterator.gen.go.tmpl]: My biggest concern with this is that it requires an interface conversion at the lowest level of the query engine. It also precludes us from changing `curs` to a `[]bufCursor` to reduce allocations. I'm ok with this change now but we might need to revisit it in the future if it's slowing down queries significantly.

**Comment by benbjohnson** [comment]
:+1:

File: tsdb/engine/tsm1/iterator.gen.go.tmpl (hunk 1 above)
Comment (jsternberg) [inline tsdb/engine/tsm1/iterator.gen.go.tmpl]: Hm, we might be able to rewrite the condition instead since, in the case of tags, the value is static. Do you want me to see how easy that would be first?

**Comment by jsternberg** [comment]
If the interface conversion causes a problem, I think we can rewrite the condition with the tag values. Doing it required more code than I was willing to rewrite at the moment, but it's definitely possible.
